File: src/bot/session-store.ts
import { mkdir, readFile, rename, writeFile } from 'node:fs/promises';
import { randomUUID } from 'node:crypto';
import { dirname } from 'node:path';
import { paths } from '../config/paths';
import { DEFAULT_BACKEND_ID, type ReasoningEffort } from '../agent/types';

/**
 * A persisted session = one Feishu topic (thread) bound to a codex thread.
 * Survives bridge restarts so @bot inside an existing topic resumes the right
 * codex thread (instead of silently starting a fresh one) and the ⚙️ per-session
 * model/effort overrides stick.
 */
export interface SessionRecord {
  /** Feishu topic thread_id (the key) */
  threadId: string;
  chatId: string;
  cwd: string;
  /** backend session id（codex 的 thread id / claude 的 session UUID）—— pass to
   * backend.resumeThread。v1 文件里的旧字段名在 read() 时迁移（见 migrate）。 */
  sessionId: string;
  /** 创建该会话的 agent 后端 id（见 src/agent/index.ts 注册表）。重启后
   * resolveThread 按它路由 resume —— 项目事后换后端不影响既有会话的归属。
   * v1 文件缺省 → 默认 codex 后端（read() 时回填）。 */
  backend: string;
  model?: string;
  effort?: ReasoningEffort;
  /** first user message excerpt, for context */
  summary: string;
  /** createTime (epoch ms) of the most recent message woven into this session —
   * the high-water mark for topic-history catch-up: the next turn only pulls
   * thread messages newer than this (see context-weave.fetchThreadContext). */
  lastSeenAt?: number;
  createdAt: number;
  updatedAt: number;
}

interface StoreFile {
  version: number;
  sessions: SessionRecord[];
}

// v2：会话 id 字段改名 sessionId + 新增 backend 字段（M-8 多后端路由）。
const FILE_VERSION = 2;

/** v1 文件的旧字段名（`codexThread` + `Id`）。拼接而非字面量，是为了让「全链改名
 * 后 grep 旧名 = 0」的判据可机械验证 —— 这里是全仓唯一还认得旧名的地方。 */
const LEGACY_V1_SESSION_FIELD = 'codexThread' + 'Id';

/** 旧记录读入迁移：v1 的旧会话 id 字段 → sessionId；缺 backend 回填默认 codex
 * 后端（v1 时代只有它）。原地落盘格式只在下次写盘时才换新（read 不回写），
 * 所以迁移必须幂等且每次 read 都跑。 */
function migrate(raw: Record<string, unknown>): SessionRecord {
  const rec = raw as unknown as SessionRecord;
  if (typeof rec.sessionId !== 'string') {
    const legacy = raw[LEGACY_V1_SESSION_FIELD];
    if (typeof legacy === 'string') (rec as { sessionId: string }).sessionId = legacy;
  }
  if (typeof rec.backend !== 'string') (rec as { backend: string }).backend = DEFAULT_BACKEND_ID;
  return rec;
}

async function read(): Promise<SessionRecord[]> {
  try {
    const text = await readFile(paths.sessionsFile, 'utf8');
    const parsed = JSON.parse(text) as Partial<StoreFile>;
    if (!Array.isArray(parsed.sessions)) return [];
    return (parsed.sessions as unknown as Record<string, unknown>[]).map(migrate);
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === 'ENOENT') return [];
    throw err;
  }
}

// 同进程内并发的「读-改-写」串行化（upsertSession/patchSession）：话题天然并行
// （semaphore 默认 10），两个话题同时落盘会基于同一旧快照算结果、后写覆盖前写——
// 其中一个话题的 sessionId 绑定静默丢失，重启后上下文蒸发。与 registry.ts 的
// 同款锁一致：配合函数式 updater，把 read+算+write 收进一个临界区。
let opChain: Promise<unknown> = Promise.resolve();
function withLock<T>(fn: () => Promise<T>): Promise<T> {
  const run = opChain.then(fn, fn);
  opChain = run.then(
    () => undefined,
    () => undefined,
  );
  return run;
}

async function write(sessions: SessionRecord[]): Promise<void> {
  await mkdir(dirname(paths.sessionsFile), { recursive: true });
  const tmp = `${paths.sessionsFile}.tmp-${process.pid}-${randomUUID()}`;
  const body: StoreFile = { version: FILE_VERSION, sessions };
  await writeFile(tmp, `${JSON.stringify(body, null, 2)}\n`, 'utf8');
  await rename(tmp, paths.sessionsFile);
}

export async function listSessions(): Promise<SessionRecord[]> {
  return read();
}

export async function getSession(threadId: string): Promise<SessionRecord | undefined> {
  return (await read()).find((s) => s.threadId === threadId);
}

/** Insert or replace a session by threadId. */
export async function upsertSession(rec: SessionRecord): Promise<void> {
  return withLock(async () => {
    const sessions = await read();
    const idx = sessions.findIndex((s) => s.threadId === rec.threadId);
    if (idx === -1) sessions.push(rec);
    else sessions[idx] = rec;
    await write(sessions);
  });
}

/** Patch fields of an existing session; no-op if it doesn't exist. `patch` 可以是
 * 对象，或一个 `(s) => patch` 函数——后者在同一临界区内基于**最新盘值**计算补丁，
 * 避免并发读-改-写丢更新。 */
export async function patchSession(
  threadId: string,
  patch:
    | Partial<Omit<SessionRecord, 'threadId'>>
    | ((s: SessionRecord) => Partial<Omit<SessionRecord, 'threadId'>>),
): Promise<void> {
  return withLock(async () => {
    const sessions = await read();
    const rec = sessions.find((s) => s.threadId === threadId);
    if (!rec) return;
    const actual = typeof patch === 'function' ? patch(rec) : patch;
    const target = rec as unknown as Record<string, unknown>;
    for (const [k, v] of Object.entries(actual)) {
      if (v !== undefined) target[k] = v;
    }
    rec.updatedAt = Date.now();
    await write(sessions);
  });
}

File: internal/bot/session_store.go
package bot

// session_store.go —— topic↔backend session 绑定（对齐 TS bot/session-store）。
// 持久化：bridge 重启后 topic 内 @bot resume 正确的 codex thread（而非新启）。

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"

	"github.com/modelzen/feishu-codex-bridge/internal/agent"
)

const sessionFileVersion = 2

// SessionRecord 一个持久化会话（Feishu topic ↔ backend session）。
type SessionRecord struct {
	ThreadID   string                `json:"threadId"`
	ChatID     string                `json:"chatId"`
	Cwd        string                `json:"cwd"`
	SessionID  string                `json:"sessionId"`
	Backend    string                `json:"backend"`
	Model      string                `json:"model,omitempty"`
	Effort     agent.ReasoningEffort `json:"effort,omitempty"`
	Summary    string                `json:"summary"`
	LastSeenAt int64                 `json:"lastSeenAt,omitempty"`
	CreatedAt  int64                 `json:"createdAt"`
	UpdatedAt  int64                 `json:"updatedAt"`
}

type sessionStoreFile struct {
	Version  int             `json:"version"`
	Sessions []SessionRecord `json:"sessions"`
}

// SessionStore topic↔session 绑定存储（path 注入 + mutex 串行化 + 原子写）。
type SessionStore struct {
	path string
	mu   sync.Mutex
}

// NewSessionStore 构造（path = config.BotSessionsFile(appID)）。
func NewSessionStore(path string) *SessionStore {
	return &SessionStore{path: path}
}

// v1→v2 迁移：codexThreadId → sessionId；缺 backend → 默认 codex。
func migrateSession(raw map[string]any) SessionRecord {
	b, _ := json.Marshal(raw)
	var rec SessionRecord
	json.Unmarshal(b, &rec)
	if rec.SessionID == "" {
		// v1 旧字段 codexThreadId（拼接避免 grep 旧名）。
		if legacy, ok := raw["codexThread"+"Id"].(string); ok {
			rec.SessionID = legacy
		}
	}
	if rec.Backend == "" {
		rec.Backend = agent.DEFAULT_BACKEND_ID
	}
	return rec
}

func (s *SessionStore) read() ([]SessionRecord, error) {
	return ListSessionsIn(s.path)
}

func (s *SessionStore) write(sessions []SessionRecord) error {
	if err := os.MkdirAll(filepath.Dir(s.path), 0o755); err != nil {
		return err
	}
	b, err := json.MarshalIndent(sessionStoreFile{Version: sessionFileVersion, Sessions: sessions}, "", "  ")
	if err != nil {
		return err
	}
	b = append(b, '\n')
	tmp := fmt.Sprintf("%s.tmp-%d", s.path, os.Getpid())
	if err := os.WriteFile(tmp, b, 0o600); err != nil {
		return err
	}
	return os.Rename(tmp, s.path)
}

// List 全部会话。
func (s *SessionStore) List() ([]SessionRecord, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.read()
}

// Get 按 threadId 查。
func (s *SessionStore) Get(threadID string) (*SessionRecord, error) {
	sessions, err := s.List()
	if err != nil {
		return nil, err
	}
	for i := range sessions {
		if sessions[i].ThreadID == threadID {
			return &sessions[i], nil
		}
	}
	return nil, nil
}

// Upsert 按 threadId 插入或替换。
func (s *SessionStore) Upsert(rec SessionRecord) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	sessions, err := s.read()
	if err != nil {
		return err
	}
	found := false
	for i := range sessions {
		if sessions[i].ThreadID == rec.ThreadID {
			sessions[i] = rec
			found = true
			break
		}
	}
	if !found {
		sessions = append(sessions, rec)
	}
	return s.write(sessions)
}

// Patch 按 threadId 修改（函数式 updater 基于最新盘值）；不存在 no-op。
func (s *SessionStore) Patch(threadID string, fn func(*SessionRecord)) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	sessions, err := s.read()
	if err != nil {
		return err
	}
	for i := range sessions {
		if sessions[i].ThreadID == threadID {
			fn(&sessions[i])
			return s.write(sessions)
		}
	}
	return nil
}

// ListSessionsIn 读指定 sessions.json（含 v1 迁移）。跨 bot 聚合用。
func ListSessionsIn(path string) ([]SessionRecord, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var raw struct {
		Sessions []map[string]any `json:"sessions"`
	}
	if err := json.Unmarshal(b, &raw); err != nil {
		return nil, err
	}
	out := make([]SessionRecord, 0, len(raw.Sessions))
	for _, r := range raw.Sessions {
		out = append(out, migrateSession(r))
	}
	return out, nil
}
